integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/AppendDataVisitor.java

```diff
@@ -20,6 +20,7 @@ public AppendDataVisitor(
 
   @Override
   public List<OpenLineage.Dataset> apply(LogicalPlan x) {
+    // Needs to cast to logical plan despite IntelliJ claiming otherwise.
```

Comment:
That's weird...

Reply:
```
/Users/mobuchowski/code/OpenLineage/integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/AppendDataVisitor.java:25: error: method applyFirst in class PlanUtils cannot be applied to given types;
        PlanUtils.applyFirst(outputVisitors, ((AppendData) x).table());
                 ^
  required: List<? extends PartialFunction<T,List<R>>>,T
  found: List<QueryPlanVisitor<? extends LogicalPlan,Dataset>>,NamedRelation
  reason: inference variable T has incompatible bounds
    equality constraints: LogicalPlan
    lower bounds: NamedRelation
  where T,R are type-variables:
    T extends Object declared in method <T,R>applyFirst(List<? extends PartialFunction<T,List<R>>>,T)
    R extends Object declared in method <T,R>applyFirst(List<? extends PartialFunction<T,List<R>>>,T)
```

Reply:
Interesting

```scala
trait NamedRelation extends LogicalPlan {
  def name: String

  // When false, the schema of input data must match the schema of this relation, during write.
  def skipSchemaResolution: Boolean = false
}
```

```bash
$ javap -classpath spark-catalyst_2.12-3.1.2.jar -c org.apache.spark.sql.catalyst.analysis.NamedRelation
Compiled from "NamedRelation.scala"
public interface org.apache.spark.sql.catalyst.analysis.NamedRelation {
  public abstract java.lang.String name();

  public static boolean skipSchemaResolution$(org.apache.spark.sql.catalyst.analysis.NamedRelation);
    Code:
       0: aload_0
       1: invokespecial #17                 // InterfaceMethod skipSchemaResolution:()Z
       4: ireturn

  public boolean skipSchemaResolution();
    Code:
       0: iconst_0
       1: ireturn

  public static void $init$(org.apache.spark.sql.catalyst.analysis.NamedRelation);
    Code:
       0: return
}
```

Even though `NamedRelation` extends `LogicalPlan`, the compiled interface has no relationship to it because `LogicalPlan` actually gets compiled as an abstract class
```bash
$ javap -classpath spark-catalyst_2.12-3.1.2.jar -c org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
Compiled from "LogicalPlan.scala"
public abstract class org.apache.spark.sql.catalyst.plans.logical.LogicalPlan extends org.apache.spark.sql.catalyst.plans.QueryPlan<org.apache.spark.sql.catalyst.plans.logical.LogicalPlan> implements org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper, org.apache.spark.sql.catalyst.plans.logical.statsEstimation.LogicalPlanStats, org.apache.spark.sql.catalyst.plans.logical.QueryPlanConstraints, org.apache.spark.internal.Logging {
```

All of the `NamedRelation` implementations actually extend `LeafNode`, which is also an abstract class, but it's conceivable that a `NamedRelation` implementation could extend a different abstract class, so you can't make `NamedRelation` extend `LeafNode` and, of course, not every `LeafNode` is a `NamedRelation`, so you can't form the hierarchy the other way. Since `NamedRelation` has absolutely no logic in it, unlike the two abstract classes, it makes sense to compile it this way.